File: roots/roots.py
from typing import Callable

def bisection(f: Callable, a: float, b: float, tolerance: float, max_iterations: int=50):
    if f(a) * f(b) >= 0:
        return None
    i = 0
    root = []
    r = (a + b) / 2
    root.append(r)
    while abs(f(r)) > tolerance and i < max_iterations:
        if f(a) * f(r) < 0:
            b = r
        else:
            a = r
        r = (a + b) / 2
        i += 1
        root.append(r)
    return root

def regula_falsi(f: Callable, a: float, b: float, tolerance: float, max_iterations: int=50):
    if f(a) * f(b) >= 0:
        return None
    i = 0
    root = []
    r = (a * f(b) - b * f(a)) / (f(b) - f(a))
    root.append(r)
    while abs(f(r)) > tolerance and i < max_iterations:
        if f(a) * f(r) < 0:
            b = r
        else:
            a = r
        r = (a * f(b) - b * f(a)) / (f(b) - f(a))
        i += 1
        root.append(r)
    return root

def secant(f: Callable, x0: float, x1: float, tolerance: float, max_iterations: int=50):
    i = 0
    root = []
    r = x1 - f(x1) * (x1 - x0) / (f(x1) - f(x0))
    root.append(r)
    while abs(f(r)) > tolerance and i < max_iterations:
        x0 = x1
        x1 = r
        r = x1 - f(x1) * (x1 - x0) / (f(x1) - f(x0))
        i += 1
        root.append(r)
    return root

def newton_raphson(f: Callable, x0: float, tolerance: float, max_iterations: int=50):
    i = 0
    root = []
    r = x0 - f(x0) * 1e-5 / (f(x0 + 1e-5) - f(x0))
    root.append(r)
    while abs(f(r)) > tolerance and i < max_iterations:
        x0 = r
        r = x0 - f(x0) * 1e-5 / (f(x0 + 1e-5) - f(x0))
        i += 1
        root.append(r)
    return root

def fixed_point(f: Callable, g: Callable, x0: float, tolerance: float, max_iterations: int=50):
    i = 0
    root = []
    x = g(x0)
    root.append(x)
    while abs(f(x0)) > tolerance and i < max_iterations:
        x0 = x
        x = g(x0)
        root.append(x)
        i += 1
    return root

File: roots/roots.m
function roots
    format long;
    f = @(x) x^2 -3;
    g = @(x) (-1/3)*x^2 + x + 1;
    disp("Bisection method:");
    bisection_root = bisection(f, 0, 2, 0.0001);
    disp("Regula Falsi method:");
    regula_falsi_root = regula_falsi(f, 0, 2, 0.0001);
    disp("Secant method:");
    secant_root = secant(f, 0, 2, 0.0001);
    disp("Newton-Raphson method:");
    newton_raphson_root = newton_raphson(f, 0, 0.0001);
    disp("Fixed-point iteration method:");
    fixed_point_root = fixed_point(f, g, 2, 0.0001); 
end

function root = bisection(f, a, b, tolerance)
    if f(a) * f(b) >= 0
        disp('The function must have different signs at the endpoints.');
        root = [];
    end
    r = (a + b)/2
    root = [r];
    while abs(f(r)) > tolerance
        if f(a) * f(r) < 0
            b = r;
        endif
        if f(b) * f(r) < 0
            a = r;
        endif
        r = (a + b)/2
        root = [root, r];
    endwhile
end

function root = regula_falsi(f, a, b, tolerance)
    if f(a) * f(b) >= 0
        disp('The function must have different signs at the endpoints.');
        root = [];
    end
    r = (a * f(b) - b * f(a)) / (f(b) - f(a))
    root = [r];
    while abs(f(r)) > tolerance
        if f(a) * f(r) < 0
            b = r;
        endif
        if f(b) * f(r) < 0
            a = r;
        endif
        r = (a * f(b) - b * f(a)) / (f(b) - f(a))
        root = [root, r];
    endwhile
end

function root = secant(f, x0, x1, tolerance)
    r = x1 - f(x1) * (x1 - x0) / (f(x1) - f(x0))
    root = [r];
    while abs(f(r)) > tolerance
        x0 = x1;
        x1 = r;
        r = x1 - f(x1) * (x1 - x0) / (f(x1) - f(x0))
        root = [root, r];
    endwhile
end

function root = newton_raphson(f, x0, tolerance)
    r = x0 - f(x0) * 1e-5 / (f(x0 + 1e-5) - f(x0))
    root = [r];
    while abs(f(r)) > tolerance
        x0 = r;
        r = x0 - f(x0) * 1e-5 / (f(x0 + 1e-5) - f(x0))
        root = [root, r];
    endwhile
end

function root = fixed_point(f, g, x0, tolerance)
    x = g(x0)
    root = [x];
    while abs(f(x0)) > tolerance
        x0 = x;
        x = g(x0)
        root = [root, x];
    endwhile
end